test/remote/gateways/remote_redsys_rest_test.rb

```diff
@@ -0,0 +1,206 @@
+require 'test_helper'
+
+class RemoteRedsysRestTest < Test::Unit::TestCase
+  def setup
+    @gateway = RedsysRestGateway.new(fixtures(:redsys_rest))
+    @amount = 100
+    @credit_card = credit_card('4548812049400004')
+    @declined_card = credit_card
+    @threeds2_credit_card = credit_card('4918019199883839')
+
+    @threeds2_credit_card_frictionless = credit_card('4548814479727229')
+    @threeds2_credit_card_alt = credit_card('4548817212493017')
+    @options = {
+      order_id: generate_order_id
+    }
+  end
+
+  def test_successful_purchase
+    response = @gateway.purchase(@amount, @credit_card, @options)
+    assert_success response
+    assert_equal 'Transaction Approved', response.message
+  end
+
+  def test_purchase_with_invalid_order_id
+    response = @gateway.purchase(@amount, @credit_card, order_id: "a%4#{generate_order_id}")
+    assert_success response
+    assert_equal 'Transaction Approved', response.message
+  end
+
+  def test_failed_purchase
+    response = @gateway.purchase(@amount, @declined_card, @options)
+    assert_failure response
+    assert_equal 'Refusal with no specific reason', response.message
+  end
+
+  def test_purchase_and_refund
+    purchase = @gateway.purchase(@amount, @credit_card, @options)
+    assert_success purchase
+    refund = @gateway.refund(@amount, purchase.authorization, @options)
+    assert_success refund
+  end
+
+  def test_purchase_and_failed_refund
+    purchase = @gateway.purchase(@amount, @credit_card, @options)
+    assert_success purchase
+    refund = @gateway.refund(@amount + 100, purchase.authorization, @options)
+    assert_failure refund
+    assert_equal 'SIS0057 ERROR', refund.message
+  end
+
+  def test_purchase_and_refund_with_currency
```

Comment:
Nitpick: but maybe this test name should be something like `failed_purchase_with_unsupported_currency` since we are testing failure and not doing a refund.

Reply:
good idea!

Reply:
Done!